test/Npgsql.Tests/Types/TypeHandlerTestBase.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System.Threading.Tasks;
+using NpgsqlTypes;
+using NUnit.Framework;
+using static Npgsql.Tests.TestUtil;
+
+namespace Npgsql.Tests.Types
+{
+    public abstract class TypeHandlerTestBase<T> : MultiplexingTestBase
+    {
+        readonly string? _typeName;
+        readonly string? _minVersion;
+
+        protected TypeHandlerTestBase(MultiplexingMode multiplexingMode, string? typeName = null, string? minVersion = null)
+            : base(multiplexingMode) => (_typeName, _minVersion) = (typeName, minVersion);
+
+        [Test]
+        [TestCaseSource("TestCases")]
+        public async Task Read(string query, T expected)
+        {
+            using var conn = await OpenConnectionAsync();
+            using var cmd = new NpgsqlCommand($"SELECT {query}", conn);
+
+            if (_minVersion is string minVersion)
+                MinimumPgVersion(conn, minVersion);
+
+            Assert.AreEqual(await cmd.ExecuteScalarAsync(), expected);
+        }
+
+        [Test]
+        [TestCaseSource("TestCases")]
+        public async Task Write(string query, T expected)
+        {
+            using var conn = await OpenConnectionAsync();
+            using var cmd = new NpgsqlCommand($"SELECT {query}::text = @p::text", conn)
+            {
+                Parameters = { new NpgsqlParameter<T>("p", expected) { DataTypeName = _typeName } }
```

Comment:
We should also test that NpgsqlDbType (and even DbType, if provided by implementing subclasses) do proper inference as well.

Reply:
There is no corresponding `DbType`, don't forget that other databases are not such powerful as PostgreSQL.

Reply:
> There is no corresponding DbType, don't forget that other databases are not such powerful as PostgreSQL.

This seems to be a generic test base class, that could work well for other types which do have a DbType. For NpgsqlDbType, there should always be one.

> Ideally we should separate mapping from handling.

I don't see why... Any testing of mapping/inference would need to send, which would go through type handling... Any reason to have two test classes where one would suffice

Reply:
While it's generic, it's hard to use it for all cases we have right now (:

> Any reason to have two test classes where one would suffice

I would like to separate mapping from handling, so handlers will know only runtime and PostgreSQL types, but from the user perspective maybe you're right... Need to think how to cleanup this part.